akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala

```diff
@@ -63,17 +63,32 @@ private[http] object StreamUtils {
 
   def captureTermination[T, Mat](source: Source[T, Mat]): (Source[T, Mat], Future[Unit]) = {
     val promise = Promise[Unit]()
-    val transformer = new PushStage[T, T] {
-      def onPush(element: T, ctx: Context[T]) = ctx.push(element)
-      override def onUpstreamFailure(cause: Throwable, ctx: Context[T]) = {
-        promise.failure(cause)
-        ctx.fail(cause)
-      }
-      override def postStop(): Unit = {
-        promise.trySuccess(())
+    val transformer = new SimpleLinearGraphStage[T] {
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
```

Comment:
btw, the logic should extend the inHandler and outHandler. Not just that it looks simpler, it also saves on allocations and two object headers.